x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinatorTests.java

```diff
@@ -2350,6 +2353,94 @@ public void testFollowerIndexIsCreatedInExecuteAutoFollow() {
         assertThat(autoFollowResults.v2().contains(indexName), equalTo(true));
     }
 
+    public void testRemovesStatsOnRemovingAutoFollowPattern() {
+        // given auto-follow pattern added
+        var pattern1 = createAutoFollowPattern("remote1", "logs-*");
+        var pattern2 = createAutoFollowPattern("remote2", "logs-*");
+        var pattern3 = createAutoFollowPattern("remote2", "metrics-*");// same remote
+
+        var autoFollowCoordinator = createAutoFollowCoordinator();
+        autoFollowCoordinator.updateAutoFollowers(
+            createClusterStateWith(Map.of("pattern1", pattern1, "pattern2", pattern2, "pattern3", pattern3))
+        );
+
+        // and stats are published
+        autoFollowCoordinator.updateStats(
+            List.of(
+                new AutoFollowCoordinator.AutoFollowResult("pattern1", new RuntimeException("ClusterStateFetchException")),
+                new AutoFollowCoordinator.AutoFollowResult(
+                    "pattern1",
+                    List.of(Tuple.tuple(new Index("logs-1", UUIDs.base64UUID()), new RuntimeException("AutoFollowExecutionException")))
+                ),
+                new AutoFollowCoordinator.AutoFollowResult("pattern2", new RuntimeException("ClusterStateFetchException")),
+                new AutoFollowCoordinator.AutoFollowResult(
+                    "pattern2",
+                    List.of(Tuple.tuple(new Index("logs-1", UUIDs.base64UUID()), new RuntimeException("AutoFollowExecutionException")))
+                ),
+                new AutoFollowCoordinator.AutoFollowResult("pattern3", new RuntimeException("ClusterStateFetchException")),
+                new AutoFollowCoordinator.AutoFollowResult(
+                    "pattern3",
+                    List.of(Tuple.tuple(new Index("metrics-1", UUIDs.base64UUID()), new RuntimeException("AutoFollowExecutionException")))
+                )
+            )
+        );
+
+        // when auto-follow pattern `pattern3` is removed
+        var before = autoFollowCoordinator.getStats();
+        autoFollowCoordinator.updateAutoFollowers(createClusterStateWith(Map.of("pattern1", pattern1, "pattern2", pattern2)));
+        autoFollowCoordinator.updateStats(List.of());// actually triggers the purge
+        var after = autoFollowCoordinator.getStats();
+
+        // then stats are removed as well (but only for the removed pattern)
+        assertThat(
+            before.getRecentAutoFollowErrors(),
+            allOf(hasKey("pattern1:logs-1"), hasKey("pattern2:logs-1"), hasKey("pattern3:metrics-1"))
```

Comment:
I think it would be simpler to read using `equalTo` on the `keySet()`? Likewise in the next assert.